### PR TITLE
Detach Reattach mechanism in the upload command

### PR DESCRIPTION
**Pull request by ChanochShayner**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/46381

## Description
Added the Detach Reattach mechanism in the upload command, for the CI/CD users.

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Review by guykeller — COMMENTED** [review]
Reminder - waiting for you to self-CR since there are some improvements here that you should be able to spot.

Some examples provided, LMK if you'd like to talk about it

File: demisto_sdk/__main__.py [hunk 1]
```diff
@@ -886,14 +893,15 @@ def upload(**kwargs):
...
+            pack_path = config_file_to_parse.get_custom_packs_paths()
+            kwargs['detached_files'] = True
             kwargs.pop('input_config_file')
 
         output_zip_path = kwargs.pop('keep_zip') or tempfile.gettempdir()
         packs_unifier = PacksZipper(pack_paths=pack_path, output=output_zip_path,
                                     content_version='0.0.0', zip_all=True, quiet_mode=True)
         packs_zip_path, pack_names = packs_unifier.zip_packs()
-        if packs_zip_path is None:
+        if packs_zip_path is None and not kwargs.get('detached_files'):
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/__main__.py:910]: Sure

File: demisto_sdk/commands/upload/uploader.py [hunk 2]
```diff
@@ -404,15 +419,152 @@ def __init__(self, config_file_path: str):
...
+        return detach_files_list
+
+    def is_valid_file_for_detach(self, file_path: str) -> bool:
+        valid_file = False
+        for file in self.VALID_FILES_FOR_DETACH:
+            if file in file_path and (file_path.endswith('yml') or file_path.endswith('json')):
+                return True
+        return valid_file
+
+    def find_item_type_to_detach(self, file_path) -> str:
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/upload/uploader.py:483]: We extract them from the repo in a specific way, not sure this functions fits other kind of files

File: demisto_sdk/commands/upload/uploader.py [hunk 3]
```diff
@@ -404,15 +419,152 @@ def __init__(self, config_file_path: str):
...
+        file_type: str
+        if 'Playbooks' in file_path or 'Scripts' in file_path:
+            file_type = 'yml'
+        else:
+            file_type = 'json'
+        return file_type
+
+    def find_item_id_to_detach(self):
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/upload/uploader.py:486]: We extract them from the repo in a specific way, not sure this function fits another kind of files.

File: demisto_sdk/__main__.py [hunk 4]
```diff
@@ -869,6 +869,13 @@ def format(
     "-v", "--verbose",
     help="Verbose output", is_flag=True
 )
+@click.option(
+    "--reattach",
+    help="Reattach the detached files in the XSOAR instance"
+         "For the CI/CD Flow (if you specify the --input-config-file flag, "
+         "all the items in the instance that detached and are not exist in the SystemPacks folder in the repo "
+         "will reattached.)", is_flag=True
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/__main__.py]: done

File: demisto_sdk/__main__.py [hunk 5]
```diff
@@ -886,14 +893,15 @@ def upload(**kwargs):
         else:
             config_file_path = kwargs['input_config_file']
             config_file_to_parse = ConfigFileParser(config_file_path=config_file_path)
-            pack_path = config_file_to_parse.parse_file()
+            pack_path = config_file_to_parse.get_custom_packs_paths()
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/__main__.py]: removed

File: demisto_sdk/commands/upload/uploader.py [hunk 6]
```diff
@@ -106,6 +107,8 @@ def __init__(self, input: str, insecure: bool = False, verbose: bool = False, pa
         self.demisto_version = get_demisto_version(self.client)
         self.pack_names = pack_names
         self.skip_upload_packs_validation = skip_validation
+        self.detached_files = detached_files
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/upload/uploader.py]: done

File: demisto_sdk/commands/upload/uploader.py [hunk 7]
```diff
@@ -404,15 +419,152 @@ def __init__(self, config_file_path: str):
...
+                if file_id:
+                    detach_files_list.append({'file_id': file_id, 'file_type': file_type, 'file_path': file_path})
+        return detach_files_list
+
+    def is_valid_file_for_detach(self, file_path: str) -> bool:
+        valid_file = False
+        for file in self.VALID_FILES_FOR_DETACH:
+            if file in file_path and (file_path.endswith('yml') or file_path.endswith('json')):
+                return True
+        return valid_file
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/upload/uploader.py]: done

File: demisto_sdk/commands/upload/uploader.py [hunk 8]
```diff
@@ -404,15 +419,152 @@ def __init__(self, config_file_path: str):
...
+
+    def find_item_type_to_detach(self, file_path) -> str:
+        file_type: str
+        if 'Playbooks' in file_path or 'Scripts' in file_path:
+            file_type = 'yml'
+        else:
+            file_type = 'json'
+        return file_type
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/upload/uploader.py]: done

**Review by guykeller — APPROVED** [review]

**Pull request merged**